unison-cli/src/Unison/Codebase/Editor/HandleInput.hs

```diff
@@ -2899,24 +2865,107 @@ loadPropagateDiffDefaultPatch inputDescription maybeDest0 dest = do
+          )
+    before <- Cli.getRootBranch0
+    description <- inputDescription inputs
+    Cli.stepManyAt description deleteTypesTerms
+    case doutput of
+      DeleteOutput'Diff -> do
+        after <- Cli.getRootBranch0
+        (ppe, diff) <- diffHelper before after
+        Cli.respondNumbered (ShowDiffAfterDeleteDefinitions ppe diff)
+      DeleteOutput'NoDiff -> do
+        Cli.respond Success
+  else do
+      ppeDecl <- currentPrettyPrintEnvDecl Backend.Within
+      let combineRefs = List.foldl (Map.unionWith NESet.union) Map.empty endangeredDeletions
+      Cli.respondNumbered (CantDeleteDefinitions ppeDecl combineRefs)
+
 -- | Goal: When deleting, we might be removing the last name of a given definition (i.e. the
 -- definition is going "extinct"). In this case we may wish to take some action or warn the
 -- user about these "endangered" definitions which would now contain unnamed references.
 getEndangeredDependents ::
-  -- | Which names we want to delete
+  -- | Single target for deletion
   Names ->
+  -- | All names we want to delete (including the target)
+  Set LabeledDependency ->
   -- | All names from the root branch
   Names ->
   -- | map from references going extinct to the set of endangered dependents
   Sqlite.Transaction (Map LabeledDependency (NESet LabeledDependency))
-getEndangeredDependents namesToDelete rootNames = do
+getEndangeredDependents targetNamesToDelete allTermsToDelete rootNames = do
+  -- names of terms left over after target deletion
   let remainingNames :: Names
-      remainingNames = rootNames `Names.difference` namesToDelete
-      refsToDelete, remainingRefs, extinct :: Set LabeledDependency
-      refsToDelete = Names.labeledReferences namesToDelete
-      remainingRefs = Names.labeledReferences remainingNames -- left over after delete
-      extinct = refsToDelete `Set.difference` remainingRefs -- deleting and not left over
-      accumulateDependents :: LabeledDependency -> Sqlite.Transaction (Map LabeledDependency (Set LabeledDependency))
+      remainingNames = rootNames `Names.difference` targetNamesToDelete
+  let refsToDelete :: Set LabeledDependency
+      refsToDelete = Names.labeledReferences targetNamesToDelete
+  -- remove the target from the set of names to delete
+  let allOtherNamesToDelete :: Set LabeledDependency
+      allOtherNamesToDelete = Set.difference allTermsToDelete refsToDelete
+  -- left over after deleting target
+  let remainingRefs :: Set LabeledDependency
+      remainingRefs =  Names.labeledReferences remainingNames
+  -- remove the other targets for deletion from the remaining terms
+  let remainingRefsWithoutOtherTargets :: Set LabeledDependency
+      remainingRefsWithoutOtherTargets = Set.difference remainingRefs allOtherNamesToDelete
+  -- deleting and not left over
+  let extinct :: Set LabeledDependency
+      extinct = refsToDelete `Set.difference` remainingRefsWithoutOtherTargets
```

Comment:
I'll admit I'm having a bit of a hard time tracking all the Set arithmetic here;

But in my potentially flawed understanding, it seems like `refsToDelete` shouldn't ever have anything in `remainingRefsWithoutOtherTargets` because the latter is ultimately created from `remainingNames` which has already subtracted `targetNamesToDelete` which is what's used to generate `refsToDelete`; does that track? Haha.

Might be nice to see if we can simplify this logic somehow 🤔

Reply:
Simplified by one subtraction! We didn't need to remove the single term/type to delete from the set of the user's desired deletions.